### ISSUE TITLE
Allow read/update/delete Prepared Query by name

### ISSUE THREAD

**Issue by engel75** [hashicorp/consul#10531]
#### Feature Description

Currently the UUID is needed to read, update or delete a prepared query. This is quite unflexible as the name has to be unique as well. So it looks like there is no reason to use force the user to use the uuid.
It is even more strange as for execute7explain the uuid AND the name are valid parameters:

https://www.consul.io/api/query

Be allowing the query name for read/update/delete as well the API will become much more user friendly.

#### Use Case(s)

deploy, debugging, ...

**Comment by jkirschner-hashicorp** [comment]
Hi @engel75,

Thank you for sharing with us this suggestion to enhance the API.

To help us think through this suggestion, can you tell us more about:

- What is the behavior you are hoping for if you provide a query name (instead of UUID) for `read/update/delete`?
  - e.g., affect/match a single query? or multiple?
  - e.g., exact match? or longest prefix match (like the `explain` and `execute` endpoints do currently)?
    - longest prefix match means that if you have a query template with name `abc`, then call the execute endpoint for name `abc-something`, the `abc` query template will be executed if queries have a longer prefix match (e.g., `abc-some`, `abc-something`)
- Can you tell me more about the scenario you encountered that inspired this suggestion? What are the steps you took to accomplish what you needed (because you couldn't read/update/delete by name)?

This information will help us better understand the context and consider approaches. Thank you for your contribution!

**Comment by jkirschner-hashicorp** [comment]
Hi @engel75 - we'd appreciate your perspective on the questions above when you have a chance. We want to make sure we understand where you were coming from with this suggestion before we consider possible approaches. Thanks again!

**Comment by engel75** [comment]
Hi @jkirschner-hashicorp,

the main problem, in my case, is, I do not know the UUID before the query was created. This is a problem if I try to manage my queries with, eg. ansible.

So how to manage those queries if the UUID is needed to read/update/delete them? As "name" is unique, why can't we use "name" to manage them?

**Comment by jkirschner-hashicorp** [comment]
It sounds like you're looking to match to single queries on an exact name match for one read/update/delete call, rather than match multiple queries at once or perform a non-exact match.

I can see how it would be difficult to manage based on the UUID in your case. Have you figured out a workaround that does what you need for now, even though it's not ideal?

You are right that a non-empty name could be used to manage queries. That said, there are a few things we need to think through:
- The `:uuid` parameter for `explain` and `endpoint` accepts a uuid or a name, but in those cases, the name field performs a longest prefix match rather than an exact match. We wouldn't want the same parameter to mean two different things depending on endpoint (longest prefix vs exact), as that can create confusion. We'd need to think about how to avoid that confusion.
- Any approach considered must ensure backwards compatibility, so existing users aren't adversely affected.

**Comment by engel75** [comment]
Ok understood. But regarding backwards compatibility I see no problem as you would just "add" not change anything. you would add the possibility to delete/update/read a query by its name. Still the UUID is valid too. But I understand your doubts regarding using name in two different manners. From my point of view it would be fine if the documentation is very detailed in such a case.

I did not find/build any workaround. Currently,  I manage them "manual" which is not good.

Any chance to have a look at too?
https://github.com/hashicorp/consul/issues/9780

**Comment by jkirschner-hashicorp** [comment]
One potential workaround I see is to use the prepared queries `list` endpoint, find the query in the response list that has the name you care about, then use the uuid from that query in subsequent calls to read/update/delete. Certainly not convenient though.

If we have thoughts on a good approach, we'll post here to see what you/others think.

I'll also make sure 9780 gets a look soon - thanks!

**Comment by engel75** [comment]
Thank you @jkirschner-hashicorp for looking into this one and your input regarding a workaround. I will give it a try as soon as I got some spare time.

**Comment by jkirschner-hashicorp** [comment]
Hello Consul community members,

We would welcome a PR contributed by the community for this enhancement! If you're interested, please comment here so anyone interested can stay informed.

The approach should ensure the following:
- The path for read, update, and delete must be `/query/exact-name/:name`
  - We decided against a design in which the path is `/query/:uuidOrName` because we didn't want to create confusion for users as to how a provided name would be used. For read/update/delete, name would require an exact match. But execute and explain use a longest prefix match. That difference is important, but could be easily overlooked if the path was `/query/:uuidOrName`.

Files of interest for implementation:
- `agent/consul/state/prepared_query.go`: where prepared query operations are currently performed
- `agent/consul/prepared_query_endpoint_test.go`: automated tests for the above
- `agent/http_register.go`: where HTTP handlers are registered

**Comment by tmnhat2001** [comment]
@jkirschner-hashicorp I would like to take on this issue if it is still needed.

**Comment by jkirschner-hashicorp** [comment]
Hi @tmnhat2001,

You're welcome to take this on!

Just to set expectations, this may be a harder task than it seems. I worked on this a little bit before posting the guidance above.  It involves some working with memdb and channels (which I had no prior experience with - I haven't used Golang myself much yet), and a grasp of the difference between a prepared query and a prepared query template.

I see you recently contributed a front-end PR! If you haven't worked with our back-end code before, take a look at the [latest contributing guide](https://github.com/hashicorp/consul/blob/864a20c59185e8d910488a9e03612d093a7b6254/.github/CONTRIBUTING.md) (which should be merged soon).

Let us know if you have any questions as you go!

**Comment by tmnhat2001** [comment]
@jkirschner-hashicorp I have some updates and a couple of questions.

I think the implementation for this could be split into at least 3 PR's. The implementation of each read, update and delete operation can be a separate PR.

Is there a guideline on the size of the PR? The implementation for the read endpoint is already at least 350 lines of code. I'm not sure whether this would be too large of a PR.

Do we need do implement the OPTIONS HTTP method for `/query/exact-name`? This method is currently implemented for `/query/` endpoints.
https://github.com/hashicorp/consul/blob/6a7d56951cac9e877fcdc376381f787f824e84c8/agent/prepared_query_endpoint.go#L317-L319

Do we need to add these new endpoints to the [api package](https://github.com/hashicorp/consul/tree/main/api) as well?

**Comment by jkirschner-hashicorp** [comment]
@tmnhat2001 : I suggest creating a PR from whatever you have so far and in its description:
- reference this issue
- say you're looking for feedback on the approach shown for the read endpoint before you proceed to the update and delete endpoints
- ask your questions there (e.g., whether they prefer 1 PR per endpoint or a combined PR)

An engineer on the team will review and give feedback. Because team members are taking time off during the winter holidays, it may take longer than usual to get a response (e.g., early January).

Answering your questions as best as I can:

> Do we need to add these new endpoints to the api package as well?

Yes, we should.

> Do we need do implement the OPTIONS HTTP method for /query/exact-name? This method is currently implemented for /query/ endpoints.

Good catch. Yes, I think it makes sense to provide the same support for the OPTIONS HTTP method as the id-based query endpoints.

**Comment by jkirschner-hashicorp** [comment]
Hi @tmnhat2001 - just checking in, is the code in a state where it would be beneficial to have an initial PR review?

**Comment by tmnhat2001** [comment]
@jkirschner-hashicorp Yes, definitely. I'll try to put up a PR around mid to late this week.

**Comment by tmnhat2001** [comment]
@jkirschner-hashicorp Just fyi, I have opened this PR for review #12053.

**Comment by kisunji** [comment]
Thanks @tmnhat2001! We'll aim to review this shortly. Sorry for the delay!

**Comment by tmnhat2001** [comment]
@kisunji No worries at all, there seems to be some failing tests, I'll look into them in the meantime.
